### PR TITLE
Support address aggregation

### PR DESCRIPTION
**Pull request by shenkeyao**
Changes include:
* `transfer` functions: A source address is now optional. When it's `None`, all addresses may be used for the transfer.
* CLI: The original `transfer` command is renamed to `transfer_from`. The current `transfer` command doesn't need a source address.
* Tests: When initiating the network, each wallet will have two addresses so that the address/balance aggregation can be tested.

Closes https://github.com/EspressoSystems/seahorse/issues/26.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -686,11 +686,28 @@ impl<L: Ledger> Default for EventSummary<L> {
 }
 
 impl<'a, L: 'static + Ledger> WalletState<'a, L> {
+    fn keypairs(&self) -> Vec<UserKeyPair> {
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/lib.rs]: Also changed other `*_keypair` to `*_key_pair` except for `FeeInput::owner_keypair` and `FreezeNoteInput::keypair` which are defined by Jellyfish.

File: src/testing/mod.rs [hunk 2]
```diff
@@ -215,14 +215,21 @@ pub trait SystemUnderTest<'a>: Default + Send + Sync {
     ) -> Self::MockNetwork;
     async fn create_storage(&mut self) -> Self::MockStorage;
 
+    /// `two_addresses_per_wallet`
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/testing/mod.rs]: Changed this to always create two addresses, but haven't added a test for the one-address API though.

File: src/lib.rs [hunk 3]
```diff
@@ -2100,50 +2122,58 @@ impl<'a, L: 'static + Ledger, Backend: 'a + WalletBackend<'a, L> + Send + Sync>
...
     /// To skip an output when generating memos, call `genearte_memos` after removing the record from
     /// the list of outputs.
+    ///
+    /// `account`
+    /// * If provided, only this address will be used to transfer the asset.
+    /// * Otherwise, all the owned addresses can be used for the transfer.
     pub async fn transfer(
         &mut self,
-        account: &UserAddress,
+        account: Option<&UserAddress>,
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/lib.rs:2234]: Sounds good. Let me create an issue for this.

[the remark this answers is not in the document]
Comment (shenkeyao) [reply src/lib.rs:2234]: Issue: https://github.com/EspressoSystems/seahorse/issues/68.

[the remark this answers is not in the document]
Comment (jbearer) [reply src/lib.rs:2234]: Oh I also created #67 haha. I'll just link them

File: src/cli.rs [hunk 4]
```diff
@@ -475,7 +475,7 @@ fn init_commands<'a, C: CLI<'a>>() -> Vec<Command<'a, C>> {
                         io,
                         "{} {}",
                         UserAddress(pub_key.address()),
-                        wallet.balance(&pub_key.address(), &asset.item).await
+                        wallet.balance_breakdown(&pub_key.address(), &asset.item).await
```
Comment (jbearer) [inline src/cli.rs:478]: At the end of this loop, we should also print something like `Total {}, wallet.balance(&asset.item).await`

File: src/lib.rs [hunk 5]
```diff
@@ -2168,7 +2195,7 @@ impl<'a, L: 'static + Ledger, Backend: 'a + WalletBackend<'a, L> + Send + Sync>
     /// Compute the balance frozen records of the given asset type owned by the given address.
     pub async fn frozen_balance(&self, account: &UserAddress, asset: &AssetCode) -> u64 {
```
Comment (jbearer) [inline src/lib.rs:2196]: Can we change this to `frozen_balance_breakdown` to be more consistent? Don't have to add `frozen_balance(self, asset)` now, but we should at least reserve the name for it.

File: src/lib.rs [hunk 6]
```diff
@@ -2619,7 +2657,7 @@ impl<'a, L: 'static + Ledger, Backend: 'a + WalletBackend<'a, L> + Send + Sync>
         } else {
             let (sender, receiver) = oneshot::channel();
 
-            if state.user_keys.contains_key(&receipt.submitter) {
+            if state.user_keys.contains_key(&receipt.submitters[0]) {
```
Comment (jbearer) [inline src/lib.rs:2660]: Above (in `transaction_status`) we consider a transaction "submitted by this wallet" if _all_ of the `submitters` are in `user_keys`. I think that is correct, since the wallet only generates transactions where all of the submitters are in the wallet, and we should try to be consistent here.

```suggestion
            if receipt.submitters.iter().all(|key| state.user_keys.contains_key(key)) {
```

File: src/txn_builder.rs [hunk 7]
```diff
@@ -718,10 +721,10 @@ pub struct TransactionHistoryEntry<L: Ledger> {
     pub time: DateTime<Local>,
     pub asset: AssetCode,
     pub kind: TransactionKind<L>,
-    // If we sent this transaction, `sender` records the address of the spending key used to submit
-    // it. If we received this transaction from someone else, we may not know who the sender is and
-    // this field may be None.
-    pub sender: Option<UserAddress>,
+    // If we sent this transaction, `senders` records the addresses of the spending keys used to
+    // submit it. If we received this transaction from someone else, we may not know who the senders
+    // are and this field may be None.
```
Comment (jbearer) [inline src/txn_builder.rs:726]: This comment is out of sync. I actually kind of like making it an `Option<Vec<UserAddress>>` and continuing to use `None` for "unknown", but i think since we never build a transaction with empty `senders` ourselves, it is also fine to use `vec![]` for "unknown", in which case just update the comment.

File: src/txn_builder.rs [hunk 8]
```diff
@@ -1117,92 +1137,128 @@ impl<L: Ledger> TransactionState<L> {
...
         }
-        let fee_input = self.find_fee_input(spec.owner_keypair, spec.fee)?;
 
         // prepare outputs, excluding fee change (which will be automatically generated)
-        let mut outputs = vec![];
+        let mut outputs = Vec::new();
         for (pub_key, amount) in spec.receivers {
             outputs.push(RecordOpening::new(
-                rng,
+                &mut rng.clone(),
```
Comment (jbearer) [inline src/txn_builder.rs:1189]: Why is this needed?

File: src/txn_builder.rs [hunk 9]
```diff
@@ -1469,37 +1539,83 @@ impl<L: Ledger> TransactionState<L> {
...
+            &vec![owner_key_pair.clone()],
+            FreezeFlag::Unfrozen,
+            fee,
+            Some(1),
+        )?[0]
+            .1
+            .clone()
+            .into_iter()
+            .next()
+            .unwrap();
```
Comment (jbearer) [inline src/txn_builder.rs:1613]: I think this is the same as `.remove(0)`

File: src/txn_builder.rs [hunk 10]
```diff
@@ -1469,37 +1539,83 @@ impl<L: Ledger> TransactionState<L> {
...
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn find_records(
+        &self,
+        asset: &AssetCode,
+        owner_key_pairs: &[UserKeyPair],
+        frozen: FreezeFlag,
+        amount: u64,
+        max_records: Option<usize>,
```
Comment (jbearer) [inline src/txn_builder.rs:1564]: This function appears to no longer respect `max_records`. Each individual call to `find_records_with_pub_key` will respect it, but overall we might return as many as `max_records*owner_key_pairs.len()` records, if I'm not mistaken. This will cause problems with `find_fee_input`

File: src/txn_builder.rs [hunk 11]
```diff
@@ -1469,37 +1539,83 @@ impl<L: Ledger> TransactionState<L> {
...
-        let (ro, uid) = self
-            .find_records(
-                &AssetCode::native(),
-                &owner_keypair.pub_key(),
-                FreezeFlag::Unfrozen,
-                fee,
-                Some(1),
-            )
-            .map(|(ros, _change)| ros.into_iter().next().unwrap())?;
+        let (ro, uid) = self.find_records(
```
Comment (jbearer) [inline src/txn_builder.rs:1602]: Should we just use `find_records_with_pub_key` since we only have one key pair?

**Review by jbearer — COMMENTED** [review]
This looks great! A few smallish comments and then I'll approve

File: src/txn_builder.rs (hunk 8 above)
Comment (shenkeyao) [reply src/txn_builder.rs]: Oh it was needed previously, but not anymore. I've reverted it.

**Review by jbearer — APPROVED** [review]
🎉
